src/factory.rs

```diff
@@ -1,37 +1,82 @@
 use crate::*;
+use std::marker::PhantomData;
+use std::sync::atomic::{AtomicPtr, Ordering};
+
+/// Attempts to load and cache the factory interface for the given WinRT class.
+pub struct FactoryCache<C: RuntimeName, I: ComInterface> {
+    pub shared: AtomicPtr<std::ffi::c_void>,
+    pub _c: PhantomData<C>,
+    pub _i: PhantomData<I>,
+}
+
+impl<C: RuntimeName, I: ComInterface + Default> FactoryCache<C, I> {
+    pub fn call<R, F: FnOnce(&I) -> Result<R>>(&mut self, callback: F) -> Result<R> {
+        unsafe {
+            loop {
+                // Attempt to load a previously cached factory pointer.
+                let ptr = self.shared.load(Ordering::Relaxed);
+
+                // If a pointer is found, the cache is primed and we're good to go.
+                if !ptr.is_null() {
+                    return callback(std::mem::transmute(&ptr));
+                }
+
+                // Otherwise, we load the factory the usual way.
+                let factory = factory::<C, I>()?;
+
+                // If the factory is agile, we can safely cache it.
+                if factory.is_agile() {
+                    if self
+                        .shared
+                        .compare_and_swap(
+                            std::ptr::null_mut(),
+                            std::mem::transmute_copy(&factory),
+                            Ordering::Relaxed,
+                        )
+                        .is_null()
+                    {
+                        std::mem::forget(factory);
+                    }
+                } else {
+                    // Otherwise, for non-agile factories we simply use the factory
+                    // and discard after use as it is not safe to cache.
+                    return callback(&factory);
+                }
+            }
+        }
+    }
+}
 
 /// Attempts to load the factory interface for the given WinRT class.
-///
-/// Note that factory caching still needs to be implemented.
-pub fn factory<C: RuntimeName, I: ComInterface>() -> Result<I> {
-    let mut factory = std::ptr::null_mut();
+pub fn factory<C: RuntimeName, I: ComInterface + Default>() -> Result<I> {
+    let mut factory = I::default();
```

Comment:
I like this change. It would be nice to eliminate as many uses of std::mem::transmute as possible since it's such a sharp tool.